### ISSUE TITLE
Consider sharing diagnostics issued by JSON source generator with reflection-based serializer

### ISSUE THREAD

**Issue by layomia** [dotnet/runtime#61262]
Many of the [diagnostics](https://github.com/dotnet/runtime/blob/1c7100b0c21a2af147362c7b5c61408611743ded/src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs#L121-L167) issued by the JSON source generator are also applicable to the reflection-based serializer. For example, a type cannot contain multiple `[JsonConstructor]` applications. When the source generator is used, it logs a compile error when a type violates this expectation. A runtime exception is thrown in the reflection-based serializer instead.

Rather than having different behavior, we could package these diagnostic examinations into an analyzer that could be shared between the source generator and the reflection serializer, potentially giving the reflection serializer less validation work to do at runtime. We'd have to consider where this analyzer would live. There are a couple options

- it could ship in the same assembly as the source generator (becoming the first analyzer to be shipped in the dotnet/runtime repo)
- it could ship with the rest of the analyzers in https://github.com/dotnet/roslyn-analyzers

The source generator would be dependent on the analyzer, so we'd need to ship the analyzer in a manner that guarantees its visibility to the source generator.

cc @dotnet/area-system-text-json @jeffhandley

**Comment by jkotas** [comment]
> it could ship in the same assembly as the source generator (becoming the first analyzer to be shipped in the dotnet/runtime repo)

The new [DllImportImportGenerator](https://github.com/dotnet/runtime/tree/main/src/libraries/System.Runtime.InteropServices/gen/DllImportGenerator) is same analyzer + source generator bundle. @jkoritzinsky Is that right?

**Comment by jkoritzinsky** [comment]
Yes that is correct. We have both analyzers and generators in the DllImportGenerator assembly in dotnet/runtime.

**Comment by layomia** [comment]
Great, will look into how this is wired up.

**Comment by eiriktsarpalis** [comment]
Moving to Future, as we won't have time to work on this.
